lib/fs/fakefs_test.go

```diff
@@ -172,3 +178,722 @@ func TestFakeFSRead(t *testing.T) {
+}
+
+func testDirNames(t *testing.T, fs Filesystem) {
+	t.Helper()
+	filenames := []string{"fOO", "Bar", "baz"}
+	for _, filename := range filenames {
+		fd, err := fs.Create("/" + filename)
+		if err != nil {
+			t.Errorf("Could not create %s: %s", filename, err)
+		}
+		fd.Close()
+	}
+
+	assertDir(t, fs, "/", filenames)
+}
+
+func assertDir(t *testing.T, fs Filesystem, directory string, filenames []string) {
+	t.Helper()
+	got, err := fs.DirNames(directory)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if path.Clean(directory) == "/" {
+		filenames = append(filenames, ".stfolder")
+	}
+	sort.Strings(filenames)
+	sort.Strings(got)
+
+	if !reflect.DeepEqual(got, filenames) {
+		t.Errorf("want %s, got %s", filenames, got)
+	}
+}
+
+func testFakeFSStatInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/Foo/aaa")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd.Close()
+
+	info, err := fs.Stat("/FOO/AAA")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = fs.Stat("/fOO/aAa"); err != nil {
+		t.Fatal(err)
+	}
+
+	if info.Name() != "AAA" {
+		t.Errorf("want AAA, got %s", info.Name())
+	}
+
+	fd1, err := fs.Open("/FOO/AAA")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if info, err = fd1.Stat(); err != nil {
+		t.Fatal(err)
+	}
+
+	fd2, err := fs.Open("Foo/aAa")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = fd2.Stat(); err != nil {
+		t.Fatal(err)
+	}
+
+	if info.Name() != "AAA" {
+		t.Errorf("want AAA, got %s", info.Name())
+	}
+
+	fd1.Close()
+	fd2.Close()
+
+	assertDir(t, fs, "/", []string{"foo"})
+	assertDir(t, fs, "/foo", []string{"aaa"})
+}
+
+func TestFakeFSFileName(t *testing.T) {
+	var testCases = []struct {
+		fs     string
+		create string
+		open   string
+	}{
+		{"/foo", "bar", "bar"},
+	}
+
+	for _, testCase := range testCases {
+		fs := newFakeFilesystem(testCase.fs)
+		if _, err := fs.Create(testCase.create); err != nil {
+			t.Fatal(err)
+		}
+
+		fd, err := fs.Open(testCase.open)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if got := fd.Name(); got != testCase.open {
+			t.Errorf("want %s, got %s", testCase.open, got)
+		}
+	}
+}
+
+func testFakeFSFileNameInsens(t *testing.T, fs Filesystem) {
+	var testCases = []struct {
+		create string
+		open   string
+	}{
+		{"BaZ", "bAz"},
+	}
+
+	for _, testCase := range testCases {
+		fd, err := fs.Create(testCase.create)
+		if err != nil {
+			t.Fatal(err)
+		}
+		fd.Close()
+
+		fd, err = fs.Open(testCase.open)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		defer fd.Close()
+
+		if got := fd.Name(); got != testCase.open {
+			t.Errorf("want %s, got %s", testCase.open, got)
+		}
+	}
+}
+
+func TestFakeFSRename(t *testing.T) {
+	fs := newFakeFilesystem("/qux")
+	if err := fs.MkdirAll("/foo/bar/baz", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/foo/bar/baz/qux")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	if err := fs.Rename("/foo/bar/baz/qux", "/foo/baz/bar/qux"); err == nil {
```

Comment:
the name could be a bit more different, it was not obvious that they were different because the difference is a single character in a 11 character directory path.